sdk/core/azure-core/inc/context.hpp

```diff
@@ -0,0 +1,199 @@
+    std::string s; // if active == 3
+    std::unique_ptr<ValueBase> p; // if active == 4
+  };
+
+public:
+  ContextValue() noexcept : active(0) {}
+  ContextValue(bool b) noexcept : active(1), b(b) {}
+  ContextValue(int i) noexcept : active(2), i(i) {}
+  ContextValue(const std::string& s) : active(3), s(s) {}
+  ContextValue(std::string&& s) noexcept : active(3), s(std::move(s)) {}
+  template <
+      class DerivedFromValueBase,
+      typename std::enable_if<std::is_convertible<DerivedFromValueBase*, ValueBase*>::value, int>::
+          type
+      = 0>
+  ContextValue(std::unique_ptr<DerivedFromValueBase>&& p) noexcept : active(4), p(std::move(p))
+  {
+  }
+
+  ContextValue(ContextValue&& other) : active(other.active)
+  {
+    switch (active)
+    {
+      case 1:
+        b = other.b;
+        break;
+      case 2:
+        i = other.i;
+        break;
+      case 3:
+        ::new (&s) std::string(std::move(other.s));
+        break;
+      case 4:
+        ::new (&p) std::unique_ptr<ValueBase>(std::move(other.p));
+        break;
+    }
+  }
+
+  ~ContextValue()
+  {
+    switch (active)
+    {
+      case 3:
+        s.~basic_string();
+        break;
+      case 4:
+        p.~unique_ptr<ValueBase>();
+        break;
+    }
+  }
+
+  ContextValue& operator=(const ContextValue& other) = delete;
+
+  template <class ExpectedType> const ExpectedType& get() const noexcept;
+
+  char alternative() const noexcept { return active; }
+};
+
+template <> inline const bool& ContextValue::get() const noexcept
+{
+  if (active != 1)
+  {
+    abort();
+  }
+  return b;
+}
+
+template <> inline const int& ContextValue::get() const noexcept
+{
+  if (active != 2)
+  {
+    abort();
+  }
+  return i;
+}
+
+template <> inline const std::string& ContextValue::get() const noexcept
+{
+  if (active != 3)
+  {
+    abort();
+  }
+  return s;
+}
+
+template <> inline const std::unique_ptr<ValueBase>& ContextValue::get() const noexcept
+{
+  if (active != 4)
+  {
+    abort();
+  }
+  return p;
+}
+
+class Context
+{
+public:
+  using time_point = std::chrono::system_clock::time_point;
+  using ContextValue = ContextValue;
+
+private:
+  struct ContextSharedState
+  {
+    std::shared_ptr<ContextSharedState> parent;
+    time_point cancelAt; // or something like this
+    std::string key;
+    ContextValue value;
+
+    explicit ContextSharedState() : cancelAt(time_point::max()) {}
+
+    explicit ContextSharedState(
+        const std::shared_ptr<ContextSharedState>& parent,
+        time_point cancelAt,
+        const std::string& key,
+        ContextValue&& value)
+        : parent(parent), cancelAt(cancelAt), key(key), value(std::move(value))
+    {
+    }
+  };
+
+  std::shared_ptr<ContextSharedState> impl_;
+
+  explicit Context(std::shared_ptr<ContextSharedState> impl) : impl_(std::move(impl)) {}
+
+public:
+  Context() : impl_(std::make_shared<ContextSharedState>()) {}
+
+  Context& operator=(const Context&) = default;
```

Comment:
If you're going to make it copy assignable it probably also should be copy constructible. I'm not convinced it needs to be copy assignable though.... I've not seen an assignment in example code anyway.